eth2/simulation/src/simulation.rs

```diff
@@ -15,37 +17,205 @@ where
+            shard_state
+                .execution_environment_states
+                .push(a.initial_state.clone())
+                .map_err(|_| Error::MaxLengthExceeded {
+                    what: format!("number of execution environment states"),
+                })?;
+
+            // Each shard should have the same # of ee states as there are EEs
+            assert_eq!(
+                shard_state.execution_environment_states.len(),
+                self.store.current_beacon_state.execution_environments.len()
+            );
+        }
+
+        let ee_index = self.store.current_beacon_state.execution_environments.len() - 1;
+        Ok(EEIndex::new(ee_index as u64))
     }
 
     /// Add a new shard block containing a list of transactions that need to be executed
     /// Execute all transactions on the appropriate shards / EEs, return ShardBlock index
-    fn create_shard_block(args: args::CreateShardBlock) -> ShardSlot {
-        unimplemented!();
+    pub fn create_shard_block(&mut self, a: args::CreateShardBlock) -> Result<ShardSlot> {
+        // Get the specified ShardState (if it exists)
+        let shard_index: usize = a.shard.into();
+        let shard_state = self
+            .store
+            .current_beacon_state
+            .shard_states
+            .get_mut(shard_index)
+            .ok_or(Error::OutOfBounds {
+                what: WhatBound::Shard,
+                index: shard_index,
+            })?;
+
+        // Create the shard block from args
+        let transactions =
+            VariableList::new(a.shard_transactions).map_err(|_| Error::MaxLengthExceeded {
+                what: format!("number of shard transactions per block"),
+            })?;
+        let shard_block = ShardBlock { transactions };
+
+        // Execute transactions and update shard state for all transactions
+        for transaction in shard_block.transactions.iter() {
+            // Get the specified EE (if it exists)
+            let ee_index: usize = transaction.ee_index.into();
+            let execution_environment = self
+                .store
+                .current_beacon_state
+                .execution_environments
+                .get(ee_index)
+                .ok_or(Error::OutOfBounds {
+                    what: WhatBound::ExecutionEnvironment,
+                    index: ee_index,
+                })?;
+
+            // Get the current EE state
+            let pre_state = shard_state
+                .execution_environment_states
+                .get(ee_index)
+                .ok_or(Error::OutOfBounds {
+                    what: WhatBound::ExecutionEnvironmentState,
+                    index: ee_index,
+                })?;
+
+            // Create a new runtime with the EE code, transaction data, and pre state root
+            let wasm_code: Vec<u8> = execution_environment.wasm_code.clone().into();
+            let data: Vec<u8> = transaction.data.clone().into();
```

Comment:
You should be able to omit the clone calls here, and the conversions to `Vec`. `VariableList<u8>` [implements `Deref<Target=[u8]>`](https://github.com/quilt/simulation/blob/b14447ba282ade9bdbf75914bc3d10b1510a88bb/eth2/utils/ssz_types/src/variable_list.rs#L155).

Reply:
Wow, very clever.  Thanks for the catch, will update to this.